### ISSUE TITLE
bug: createPublicClient with webSocket transport not working properly

### ISSUE THREAD

**Issue by yassinekachrad** [wagmi-dev/viem#341]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Package Version

0.2.12

### Current Behavior

The following code does not compile : 
```ts
import { createPublicClient, PublicClient, webSocket } from "viem";
import { mainnet } from 'viem/chains'

const client: PublicClient = createPublicClient({
    chain: mainnet,
    transport: webSocket(),
});
```
Because apparently `createPublicClient`'s return value is not assignable to type `PublicClient`.

### Expected Behavior

I obviously expect the code above to compile and run.

### Steps To Reproduce

```ts
import { createPublicClient, PublicClient, webSocket } from "viem";
import { mainnet } from 'viem/chains'

const client: PublicClient = createPublicClient({
    chain: mainnet,
    transport: webSocket(),
});
```

### Anything else?

This problem does not appear when using http as transport.
Additionally, it used to work on version 0.1.22.

**Comment by fubhy** [comment]
This has "regressed" since 0.1.22 because of differences in the function parameters of some methods on the public client related to `polling` options. 

In its current form though, this is the expected behavior because you are explicitly assigning the type annotation `PublicClient` which is a generic type of the shape:

```typescript
PublicClient<TTransport, TChain, TIncludeActions>
```

If you do not specify those generic type variables (e.g. `TTransport`, they'll use their defaults, which, in case of `TTransport` is the default `Transport` base type. In this case, the inferred types for some methods on the client (watchBlocks, watchPendingTransactions, ...), do not match the actual type signatures of those functions when using the `webSocket` transport. Those are special cased e.g. here: https://github.com/wagmi-dev/viem/blob/main/src/actions/public/watchBlocks.ts#L37

To solve this, you likely don't want to assign the type annotation to `PublicClient`. By doing that, you are incorrectly narrowing the returned type of the `createPublicClient` function call. Instead, your type is actually `PublicClient<WebSocketTransport>` or, more specifically even `PublicClient<WebSocketTransport, typeof mainnet, true>`. Ideally, you'd not assign that type annotation at all though and instead have it inferred automatically. If that's not possible (there are valid reasons sometimes), you can solve your problem by assigning the proper, specifc type.

TL;DR: Fix it like this:

```typescript
const client = createPublicClient({
    chain: mainnet,
    transport: webSocket(),
});
```

Or like this:

```typescript
const client: PublicClient<WebSocketTransport> = createPublicClient({
    chain: mainnet,
    transport: webSocket(),
});
```

---

@tmm @jxom I think the tight coupling for specific transports can be solved differently. Imho, the generic code for e.g. `watchBlocks` should not have knowledge about specific transports. So instead of https://github.com/wagmi-dev/viem/blob/main/src/actions/public/watchBlocks.ts#L37 , maybe we can infer the transport specific options from the transport type variable (carry it on the transport type).

**Comment by yassinekachrad** [comment]
Yep, that fixed it! Thank you.

But yeah, in this situation, it doesn't seem right that the inferred types are not the "correct" ones, it'd be a shame to always have to make the distinction between `PublicClient<Transport>` and `PublicClient<WebSocketTransport>`

**Comment by fubhy** [comment]
> Author

That's not what I said! :)

The inferred type is correct. It simply clashes with the type that you have assigned yourself. The inferred type results in:

```typescript
type PublicClient<WebSocketTransport>
```

And you assigned it (manually), to:

```typescript
type PublicClient<Transport>
```

**Comment by yassinekachrad** [comment]
Oops, I didn't express myself clearly haha

By "correct", I meant `PublicClient` (just my interpretation of correct : _as generic as possible_ here), because it's a shame that `createPublicClient` doesn't always return an instance of type `PublicClient` (and has to be "less" generic when the chosen transport is `webSocket`). But for now, I'll just replace my `PublicClient` instances to `PublicClient<WebSocketTransport>`, that'll do just fine :)

**Comment by fubhy** [comment]
It does return a `PublicClient`. Just a more specific version of it that more closely fits what you passed to `createPublicClient`. That's exactly the beauty of typescript and you *want* it to return that specific version. If it returned a generic instance, then you would not be able to use it the way it was meant to be used (e.g. the parameters for `watchPendingTransaction` now wouldn't fit those supported by the `webSocket` transport).

Generally I'd recommend to avoid assigning type annotations explicity. If you can, I'd recommend to simply do:

```typescript
const client = createPublicClient({
    chain: mainnet,
    transport: webSocket(),
});
```

**Comment by yassinekachrad** [comment]
Ohh, I think I see the idea, thank you!
The thing is that I used to do that (but mostly out of laziness, I must admit), and since the update, I thought I could fix it by explicitely specifying the type (which was quite stupid in hindsight).

However, on another note, I just realized I can't call `getContract` with this client (with webSocket transport).

Should I open another issue or post a code snippet here to reproduce the issue?

**Comment by fubhy** [comment]
Yes please. That sounds like a big indeed
